integration-tests/src/tests/client/features/multinode_test_loop_example.rs

```diff
@@ -0,0 +1,350 @@
+use near_client::{Client, SyncAdapter, SyncMessage};
+use near_epoch_manager::shard_tracker::{ShardTracker, TrackedConfig};
+use near_epoch_manager::EpochManager;
+use near_network::client::{
+    BlockApproval, BlockResponse, ClientSenderForNetwork, ClientSenderForNetworkMessage,
+};
+use near_network::test_loop::SupportsRoutingLookup;
+use near_network::types::{
+    NetworkRequests, PeerManagerMessageRequest, PeerManagerMessageResponse, SetChainInfo,
+};
+use near_primitives::network::PeerId;
+use near_primitives::shard_layout::ShardLayout;
+use near_primitives::state_record::StateRecord;
+use near_primitives::test_utils::{create_test_signer, create_user_test_signer};
+use near_primitives::types::{AccountId, AccountInfo};
+use near_primitives::version::PROTOCOL_VERSION;
+use near_primitives_core::account::{AccessKey, Account};
+use near_primitives_core::hash::CryptoHash;
+use near_store::genesis::initialize_genesis_state;
+use near_store::test_utils::create_test_store;
+use nearcore::NightshadeRuntime;
+use std::path::Path;
+use std::sync::{Arc, RwLock};
+
+#[derive(derive_more::AsMut, derive_more::AsRef)]
+struct TestData {
+    pub dummy: (),
+    pub account: AccountId,
+    pub client: ClientActions,
+    pub sync_jobs: SyncJobsActions,
+    pub shards_manager: ShardsManager,
+}
+
+impl AsMut<TestData> for TestData {
+    fn as_mut(&mut self) -> &mut Self {
+        self
+    }
+}
+
+#[derive(EnumTryInto, Debug, EnumFrom)]
+#[allow(clippy::large_enum_variant)]
+enum TestEvent {
+    Task(Arc<TestLoopTask>),
+    Adhoc(AdhocEvent<TestData>),
+    ClientDelayedActions(TestLoopDelayedActionEvent<ClientActions>),
+    ClientEventFromNetwork(ClientSenderForNetworkMessage),
+    ClientEventFromClient(ClientSenderForClientMessage),
+    ClientEventFromSyncJobs(ClientSenderForSyncJobsMessage),
+    ClientEventFromShardsManager(ShardsManagerResponse),
+    SyncJobsEventFromClient(SyncJobsSenderForClientMessage),
+    SyncJobsEventFromSyncJobs(SyncJobsSenderForSyncJobsMessage),
+    ShardsManagerRequestFromClient(ShardsManagerRequestFromClient),
+    ClientEventFromStateSyncAdapter(SyncMessage),
+    NetworkMessage(PeerManagerMessageRequest),
+    NetworkMessageForResult(
+        MessageWithCallback<PeerManagerMessageRequest, PeerManagerMessageResponse>,
+    ),
+    SetChainInfo(SetChainInfo),
+}
+
+const ONE_NEAR: u128 = 1_000_000_000_000_000_000_000_000;
+
+// TODO(robin-near): Complete this test so that it will actually run a chain.
+// TODO(robin-near): Make this a multi-node test.
+// TODO(robin-near): Make the network layer send messages.
+#[test]
+fn test_client_with_multi_test_loop() {
+    const NUM_CLIENTS: usize = 4;
+    let builder = TestLoopBuilder::<(usize, TestEvent)>::new();
+
+    let validator_stake = 1000000 * ONE_NEAR;
+    let initial_balance = 10000 * ONE_NEAR;
+    let accounts =
+        (0..100).map(|i| format!("account{}", i).parse().unwrap()).collect::<Vec<AccountId>>();
+
+    // TODO: Make some builder for genesis.
```

Comment:
fyi, had been in talks with @pugachAG as well, and it seems like there isn't an urgent requirement for a genesis builder. But in any case, if any of us three (or anyone else) starts working on this, would be nice to inform everyone. I wouldn't like duplication of effort :)